File: submit/lab2/exercises/calc/calc.py
#!/usr/bin/env python3

import re
import sys
from collections import namedtuple


"""
program
  : expr ';' program
  | #empty
  ;
expr
  : term ( ( '+' | '-' ) term )*
  ;
term
  : '-' term
  | factor
  ;
factor
  : INT
  | '(' expr ')'
  ;
"""

#use nested function for encapsulation.
def parse(text):

    def peek(kind): return lookahead.kind == kind
    def consume(kind):
        nonlocal lookahead
        if (lookahead.kind == kind):
            lookahead = nextToken()
        else:
            print(f'expecting {kind} at {lookahead.lexeme}',
                  file=sys.stderr)
            sys.exit(1)
    def nextToken():
        nonlocal index
        if (index >= len(tokens)):
            return Token('EOF', '<EOF>')
        else:
            tok = tokens[index]
            index += 1
            return tok

    def program():
        values = []
        while (not peek('EOF')):
            values.append(expr())
            consume(';')
        return values

    def expr():
        t = term()
        while (peek('+') or (peek('-'))):
            kind = lookahead.kind
            consume(kind)
            t1 = term()
            t += (t1 if (kind == '+') else -t1)
        return t

    def term():
        if (peek('-')):
            consume('-')
            return - term()
        else:
            return factor()

    def factor():
        if (peek('INT')):
            value = int(lookahead.lexeme)
            consume('INT')
            return value
        else:
            consume('(')
            value = expr()
            consume(')')
            return value

    #begin parse()
    tokens = scan(text)
    index = 0
    lookahead = nextToken()
    value = program()
    if (not peek('EOF')):
        print(f'expecting <EOF>, got {lookahead.lexeme}', file=sys.stderr)
        sys.exit(1)
    return value

def scan(text):
    SPACE_RE = re.compile(r'\s+')
    INT_RE = re.compile(r'\d+')
    CHAR_RE = re.compile(r'.')
    def next_match(text):
        m = SPACE_RE.match(text)
        if (m): return (m, None)
        m = INT_RE.match(text)
        if (m): return (m, 'INT')
        m = CHAR_RE.match(text)  #must be last: match any char
        if (m): return (m, m.group())

    tokens = []
    while (len(text) > 0):
        (match, kind) = next_match(text)
        lexeme = match.group()
        if (kind): tokens.append(Token(kind, lexeme))
        text = text[len(lexeme):]
    return tokens

def main():
    if (len(sys.argv) != 2): usage();
    contents = readFile(sys.argv[1]);
    print(parse(contents))

def readFile(path):
    with open(path, 'r') as file:
        content = file.read()
    return content

Token = namedtuple('Token', ['kind', 'lexeme'])

def usage():
    print(f'usage: {sys.argv[0]} DATA_FILE')
    sys.exit(1)

if __name__ == "__main__":
    main()

File: submit/lab2/exercises/calc/calc.mjs
#!/usr/bin/env node

import fs from 'fs';
import Path from 'path';


/* 
program
  : expr ';' program
  | #empty
  ;
expr
  : term ( ( '+' | '-' ) term )*
  ;
term
  : '-' term
  | factor
  ;
factor
  : INT
  | '(' expr ')'
  ;
*/

function parse(text) {
  const tokens = scan(text);
  let index = 0;
  let lookahead = nextToken();
  const value = program();
  return value;
  
  function peek(kind) { return lookahead.kind === kind; }
  function consume(kind) {
    if (peek(kind)) {
      lookahead = nextToken();
    }
    else {
      console.error(`expecting ${kind} at ${lookahead.kind}`);
      process.exit();
    }
  }
  function nextToken() {
    return (
      (index >=  tokens.length) ? new Token('EOF', '<EOF>') : tokens[index++]
    );
  }

  function program() {
    const values = [];
    while (!peek('EOF')) {
      values.push(expr());
      consume(';');
    }
    return values;
  }

  function expr() {
    let t = term();
    while (peek('+') || peek('-')) {
      const kind = lookahead.kind;
      consume(kind);   
      const t1 = term();
      t += (kind === '+') ? t1 : -t1;
    }
    return t;
  }

  function term() {
    if (peek('-')) {
      consume('-');
      return - term();
    }
    else {
      return factor();
    }
  }

  function factor() {
    if (peek('INT')) {
      const value = parseInt(lookahead.lexeme);
      consume('INT');
      return value;
    }
    else {
      consume('(');
      const value = expr();
      consume(')');
      return value;
    }
  }
}


function scan(text) {
  const tokens = [];
  while (text.length > 0) {
    let m;
    if ((m = text.match(/^\s+/))) {
    }
    else if ((m = text.match(/^\d+/))) {
      tokens.push(new Token('INT', m[0]));
    }
    else {
      m = text.match(/^./);
      tokens.push(new Token(m[0], m[0]));
    }
    text = text.substring(m[0].length);
  }
  return tokens;
}


const CHAR_SET = 'utf8';
function main() {
  if (process.argv.length !== 3) usage();
  const file = process.argv[2];
  const text = fs.readFileSync(file === '-' ? 0 : file, CHAR_SET);
  const value = parse(text);
  console.log(value);
}

function usage() {
  const prog = Path.basename(process.argv[1])
  console.error(`usage: ${prog} INPUT_FILE`);
  process.exit(1);
}

class Token {
  constructor(kind, lexeme) {
    Object.assign(this, {kind, lexeme});
  }
}

main();

